[Tabular] Support different random seeds per fold

LennartPurucker opened the pull request:
This PR ensures that each fold gets a different random seed. I implemented it as part of the kwargs to avoid users setting random seeds as a hyperparameter. 

> By default, we keep a static seed of 0 to ensure reproducibility. When using a bagged model, a static seed of 0 to n_splits-1 is used to ensure that each fold has a different seed that is consistent across different models.

Note, one can disable this behavior and use a static seed across all folds (similar to the prior implementation via):


TODO:

- [x] Add code to support this for all other models. 

--- 
One can control the seed behavior via `ag_args_ensemble`:

```python
# Enable behavior of varying the seeds across folds
predictor = TabularPredictor(label="class", path=str("./ag_path")).fit(
    train_data=train_data,
    hyperparameters={'GBM':{"num_boost_round": 10}},
    ag_args_ensemble={"fold_fitting_strategy": 'sequential_local', "vary_seed_across_folds": True}
)

# Set your own random seed for all models
predictor = TabularPredictor(label="class", path=str("./ag_path")).fit(
    train_data=train_data,
    hyperparameters={'GBM':{"num_boost_round": 10}},
    ag_args_ensemble={"fold_fitting_strategy": 'sequential_local', "model_random_seed": 42}
)

# Set your own random seed per configuration
predictor = TabularPredictor(label="class", path=str("./ag_path")).fit(
    train_data=train_data,
    hyperparameters={'GBM':{"num_boost_round": 10, "ag_args_ensemble": {"model_random_seed": 42}}},
)

```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Innixma commented on core/src/autogluon/core/models/ensemble/bagged_ensemble_model.py:
```diff
@@ -60,6 +60,8 @@ class BaggedEnsembleModel(AbstractModel):
     """
 
     _oof_filename = "oof.pkl"
+    _vary_seed_across_folds: bool = True
```
can be an `ag_args_fit` param to `BaggedEnsembleModel` with default True.

this can then be specified via `ag_args_ensemble` by the user.

Innixma commented on tabular/src/autogluon/tabular/models/lgb/lgb_model.py:
```diff
@@ -285,8 +284,16 @@ def _fit(self, X, y, X_val=None, y_val=None, time_limit=None, num_gpus=0, num_cp
+        if "seed_value" in params:
+            logger.log(
+                40,
+                "\t'seed_value' found in hyperparamters is ignored!"
```
nit: hyperparameters

LennartPurucker commented:
@Innixma I have reworked the PR based on your feedback. Again, with some larger design choices that need your feedback. 

Moreover, I have added a test case and am now extending it to all other models in AG

Innixma commented on core/src/autogluon/core/models/ensemble/bagged_ensemble_model.py:
```diff
@@ -108,6 +108,8 @@ def _set_default_params(self):
             "stratify": "auto",
             "bin": "auto",
             "n_bins": None,
+            "vary_seed_across_folds": True, # If True, the seed used for each fold will be varied across folds.
```
For this PR, lets set it to False so we enable the ability to use the logic, but make it so the logic doesn't yet change AutoGluon's default behavior until we benchmark things and decide if it is beneficial.

Innixma commented on core/src/autogluon/core/models/ensemble/bagged_ensemble_model.py:
```diff
@@ -938,9 +945,13 @@ def _generate_fold_configs(
                 )
 
                 fold_fit_args_list.append(fold_ctx)
+                if vary_seed_across_folds:
+                    random_seed_value += 1
```
This is technically incorrect in the scenario where `n_repeat_start` and `fold_in_set_start` don't equal 0, such as when fitting an extra repeat of the bag in a future fit call.

The way to do it correctly would be to ensure the Nth child model gets a random_seed_value of `n-1`. Code will be a bit more complicated, but pretty sure there is a smart way to do it, as I've already had to deal with this same problem in the past somewhere in the bagged ensemble code.

Innixma commented on tabular/src/autogluon/tabular/models/lgb/lgb_model.py:
```diff
@@ -128,7 +128,12 @@ def _estimate_memory_usage_static(
 
-    def _fit(self, X, y, X_val=None, y_val=None, time_limit=None, num_gpus=0, num_cpus=0, sample_weight=None, sample_weight_val=None, verbosity=2, **kwargs):
+    def _get_random_seed_from_hyperparameters(self, hyperparameters: dict | None = None) -> tuple[bool, int | None]:
+        if "seed_value" in hyperparameters:
+            return True, hyperparameters["seed_value"]
```
Is it only `seed_value`? Unsure why, but the AG code also uses `seed` as in:

```python
        if seed_val is not None:
            train_params["params"]["seed"] = seed_val
```

Maybe I wasn't comprehensive when I originally coded it.

LennartPurucker replied:
Fixing both is needed for correct backward compatibility.
Here, seed_val was overwritting seed

Innixma commented on core/src/autogluon/core/models/ensemble/bagged_ensemble_model.py:
```diff
@@ -800,6 +802,8 @@ def _fit_folds(
             k_fold_end=k_fold_end,
             n_repeat_start=n_repeat_start,
             n_repeat_end=n_repeats,
+            vary_seed_across_folds=self.params["vary_seed_across_folds"],
+            random_seed_start_value=self.params["model_random_seed"],
```
So currently BaggedEnsembleModel ignores `random_seed` in its own `_fit` call. This seems strange, as that means that nested `BaggedEnsembleModel` would not handle `random_seed` properly, as whatever is specified by the parent `BaggedEnsembleModel` would be ignored by the children `BaggedEnsembleModel`.

Maybe we don't care enough to address this gap, but good to document it if we decide not to fix this.

LennartPurucker replied:
This case should be supported now as well

LennartPurucker commented on tabular/src/autogluon/tabular/models/lgb/lgb_model.py:
```diff
@@ -285,8 +289,10 @@ def _fit(self, X, y, X_val=None, y_val=None, time_limit=None, num_gpus=0, num_cp
-        if seed_val is not None:
-            train_params["params"]["seed"] = seed_val
+
+        self.init_random_seed(random_seed=random_seed, hyperparameters=params)
```
Okay, but do we have the correct hyperparameters at that time? 
I was not sure but thought that some model might only get them in the `_fit` call.

Innixma replied:
correct hyperparameter values are determined in the `self._init_params()` call, as that is when they are set

LennartPurucker commented on core/src/autogluon/core/models/ensemble/bagged_ensemble_model.py:
```diff
@@ -109,12 +109,15 @@ def _set_default_params(self):
             self._set_default_param_value(param, val)
         super()._set_default_params()
 
+    def _get_random_seed_from_hyperparameters(self, hyperparameters: dict | None = None) -> int | None | str:
+        return hyperparameters.get("model_random_seed", "N/A")
```
Ah nvm, should only be called if it is not None, fixing tpes...

LennartPurucker commented:
The current version should now be close to the final version

LennartPurucker commented:
Should be fixed now

LennartPurucker commented:
The latest version now also supports child_oof and refit_folds to have different seeds if one passed `model_random_seed` to `ag_args_ensemble`.

Innixma commented on tabular/src/autogluon/tabular/models/tabular_nn/torch/tabular_nn_torch.py:
```diff
@@ -164,6 +164,10 @@ def _prepare_params(params):
+    def _get_random_seed_from_hyperparameters(self, hyperparameters: dict) -> int | None | str:
+        return hyperparameters.get("seed_value", "N/A")
+
+
```
```suggestion
```
